proposals/p2360.md

```diff
@@ -0,0 +1,553 @@
+
+  // ✅ OK, same as `(Base as Iface).Static()`.
+  b.IfaceStatic();
+  Base.IfaceStatic();
+  // ❌ Uses `Base as Iface`, but method call is missing an instance.
+  Base.IfaceMethod();
+  // ✅ OK, uses `Base as Iface`.
+  b.IfaceMethod();
+  b.(Base.IfaceMethod)();
+  b.((Base as Iface).Method)();
+  b.(Iface.Method)();
+}
+
+class Derived extends Base {}
+external impl Derived as Iface { ... }
+
+fn H(d: Derived) {
+  // ✅ OK, calls `Base.Static`.
+  d.Static();
+  Derived.Static();
+  // ❌ Uses `Derived as Iface`, but method call is missing an instance.
+  Derived.Method();
+  // ✅ OK, calls `Base.Method`.
+  d.Method();
+  d.(Base.Method)();
+  d.(Derived.Method)();
+
+  // ✅ OK, same as `(Derived as Iface).Static()`.
+  d.IfaceStatic();
+  Derived.IfaceStatic();
+  // ❌ Uses `Derived as Iface`, but method call is missing an instance.
+  Derived.IfaceMethod();
+  // ✅ OK, uses `Derived as Iface`.
+  d.IfaceMethod();
+  d.(Derived.IfaceMethod)();
+  d.((Derived as Iface).Method)();
+  d.(Iface.Method)();
+  // ✅ OK, uses `Base as Iface`.
+  d.(Base.IfaceMethod)();
+  d.((Base as Iface).Method)();
+}
+```
+
+### Tuple and struct literals
+
+A tuple literal such as `(1, 2, 3)` produces a tuple value. Its type cannot be
+written directly as a literal: the literal `(i32, i32, i32)` also produces a
+tuple value, not a type value. However, the type of `(1, 2, 3)` is easy to
+express: it is the result of implicitly converting `(i32, i32, i32)` to a type,
+so for example `(i32, i32, i32) as type` evaluates to the type of `(1, 2, 3)`.
+
+There is no conversion from a tuple type to a tuple value, because tuple types
+are values of type `type`, and the type `type` does not implicitly convert to a
+tuple type. For metaprogramming, it is likely that we will benefit from having a
+mechanism to convert a tuple type into a tuple of types, but this is likely to
+be possible using a variadic:
+
+```
+// Takes a tuple of values. Returns a tuple containing the types of those values.
+fn TupleTypeToTupleOfTypes[Types:! type,...](x: (Types,...)) -> auto {
+  return (Types,...);
+}
+```
+
+### Name lookup into values with constrained types
+
+Given a generic function with a generic type parameter, uses of that parameter
+will be implicitly converted to type `type`. For example:
+
+```
+fn F[T:! Printable](x: T) { x.Print(); }
+```
+
+... is equivalent to ...
+
+```
+fn F[T:! Printable](x: T as type) { x.Print(); }
+```
+
+As a result, we can no longer describe lookup for `x` as looking into the
+type-of-type, that is, the type of the expression after `x:`, because after
+implicit conversion, that expression is of type `type`. Instead, lookup will
+look into the type of `x`, which symbolically evaluates to the value
+corresponding to `T` in type `type`, and will look at that value to determine
+where else to look. Because that value is symbolically the name of a generic
+type parameter, we will look in the type of `T`, which is `Printable`.
+
+## Rationale
```

Comment:
Given the concerns I've heard around teachability, I think its worth explicitly adding a section to try to address that.

In the section, I think we should both acknowledge that the details of this model are likely to be challenging for new-comers, even ones with experience in C++. Then I think we should cover two points at least:

1) Some ideas of how we can try to mitigate the challenge here, maybe suggestions of our initial ideas around how to approach explaining or teaching it. Not looking for a lot of detail, but just to capture some of what has already been discussed like framing the initial message around "you write a tuple type as `(i32, i32)`", and only later diving into how that works at this level of detail. In discussing with someone else, analogies came up of classic cases where we do this in physics such as "spherical cows" or "ignoring air resistance".

2) Capture the point that this is a relatively small change in practice from where Carbon's design already is today, and we can and should still be open to more significant changes if this problem ends up being severe and needing it.

Does that make sense?

I know its new content, let me know if I can help with drafting.